scipy/optimize/minpack.py

```diff
@@ -510,28 +523,34 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         values will all be 1 (if the number of parameters for the function
         can be determined using introspection, otherwise a ValueError
         is raised).
-    sigma : None or M-length sequence, optional
-        If not None, the uncertainties in the ydata array. These are used as
-        weights in the least-squares problem
-        i.e. minimising ``np.sum( ((f(xdata, *popt) - ydata) / sigma)**2 )``
-        If None, the uncertainties are assumed to be 1.
+    sigma : None or M-length sequence or MxM array, optional
+        Determines the uncertainty in `ydata`.
+        If we define measurement errors as
+        ``e = ydata - f(xdata, *popt)``, then
+
+            - If `sigma` is 1-d, it determines standard deviations of errors `e`
+              and the optimized function is ``sum((e / sigma) ** 2)``.
+            - If `sigma` is 2-d, it determines the covariance matrix of errors `e`
+              and the optimized function is ``e.T.dot(inv(sigma)).dot(e)``.
+
+        None (default) is equivalent of 1-d `sigma` filled with ones.
     absolute_sigma : bool, optional
-        If False, `sigma` denotes relative weights of the data points.
-        The returned covariance matrix `pcov` is based on *estimated*
-        errors in the data, and is not affected by the overall
-        magnitude of the values in `sigma`. Only the relative
-        magnitudes of the `sigma` values matter.
-
-        If True, `sigma` describes one standard deviation errors of
-        the input data points. The estimated covariance in `pcov` is
-        based on these values.
+        If False, `sigma` is used in a relative sense. The returned parameter
```

Comment:
> No I do not think so. \Sum (e/sigma)**2 are error-weighted residuals -- I won't call it scaled errors at all. I think in your terminology "e" is errors, while I will call them residuals and sigma as errors.

Hmm, it's confusing. Let's not call `sigma` as errors? I think it's just not correct. I suggest to use "residuals" for `y - f(y, p)` and "covariance" (of assumed measurement errors)  for `sigma`.

So in your new version I like "The returned parameter covariance matrix `pcov` is based on scaling all errors by a constant", I would rephrase it slightly "The returned covariance matrix of the estimated parameters `pcov` is based on scaling `sigma` by a constant factor". Are you OK with that?

Then "This constant is set by demanding that the reduced chi-squared is equal to unity when using this scaled `sigma`". And I would add (or replace it by): "In other words, `sigma` is scaled to match the sample variance of the residuals after the fit." (but this is optional and perhaps not necessary).